### PR TITLE
Encrypt metadata spec and types

### PR DESCRIPTION
**Pull request by paweljakubas**
- [x] Adding spec and change swagger accordingly
- [x] Add field `encrypt_metadata` to `ApiConstructTransactionPostData`
- [x] Introduce ApiDecodeTransactionPostData`
- [x] deal with all code adjustment
- [x] adjust unit tests   

### Issue Number

ADP-322

### CODE REVIEW

File: specifications/api/metadata-encrypt.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
+# Specification: Encrypting and decrypting metadata
+
+This document specifies those aspects of the "Transactions New > Construct" HTTP endpoint that relate to publishing of encrypted metadata.
+In addition "Transactions New > Decode" HTTP endpoint is described in the context of decrypting the metadata.
+
+## Metadata encryptiion
```
Comment (Unisay) [inline specifications/api/metadata-encrypt.md]:
```suggestion
## Metadata encryption
```

Comment (paweljakubas) [reply specifications/api/metadata-encrypt.md]: thanks, corrected

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 2]
```diff
@@ -3269,9 +3280,11 @@ decodeSharedTransaction
     :: forall n . HasSNetworkId n
     => ApiLayer (SharedState n SharedKey)
     -> ApiT WalletId
-    -> ApiSerialisedTransaction
+    -> ApiDecodeTransactionPostData
     -> Handler (ApiDecodedTransaction n)
-decodeSharedTransaction ctx (ApiT wid) (ApiSerialisedTransaction (ApiT sealed) _) = do
+decodeSharedTransaction ctx (ApiT wid) postData = do
+    let (ApiDecodeTransactionPostData (ApiT sealed) decryptMetadata ) = postData
```
Comment (Unisay) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]:
```suggestion
    let ApiDecodeTransactionPostData (ApiT sealed) decryptMetadata = postData
```

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 3]
```diff
@@ -3430,10 +3443,12 @@ decodeTransaction
        )
     => ApiLayer s
     -> ApiT WalletId
-    -> ApiSerialisedTransaction
+    -> ApiDecodeTransactionPostData
     -> Handler (ApiDecodedTransaction n)
 decodeTransaction
-    ctx@ApiLayer{..} (ApiT wid) (ApiSerialisedTransaction (ApiT sealed) _) = do
+    ctx@ApiLayer{..} (ApiT wid) postData = do
+    let (ApiDecodeTransactionPostData (ApiT sealed) decryptMetadata ) = postData
```
Comment (Unisay) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]:
```suggestion
    let ApiDecodeTransactionPostData (ApiT sealed) decryptMetadata = postData
```

File: lib/wallet/api/http/Cardano/Wallet/Api/Types.hs [hunk 4]
```diff
@@ -1198,12 +1202,19 @@ data ApiMultiDelegationAction
     deriving (Eq, Generic, Show)
     deriving anyclass NFData
 
+newtype ApiEncryptMetadata = ApiEncryptMetadata
+    { passphrase :: (ApiT (Passphrase "lenient")) }
```
Comment (Unisay) [inline lib/wallet/api/http/Cardano/Wallet/Api/Types.hs]:
```suggestion
    { passphrase :: ApiT (Passphrase "lenient") }
```

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs (hunk 2 above)
Comment (paweljakubas) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]: done

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs (hunk 3 above)
Comment (paweljakubas) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]: done

File: lib/wallet/api/http/Cardano/Wallet/Api/Types.hs (hunk 4 above)
Comment (paweljakubas) [reply lib/wallet/api/http/Cardano/Wallet/Api/Types.hs]: done

File: specifications/api/metadata-encrypt.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+      "metadata": "raw metadata"
+    ...
+    }
+    ```
+
+    As a result we get transaction with metadata encrypted:
+    ```
+    {
+    ...
+      "metadata": "metadata encrypted"
```
Comment (Unisay) [inline specifications/api/metadata-encrypt.md:41]: If "encrypt_metadata" is an object for extensibility purposes (an object allows to specify an encryption algorithm that was chosen), would it also make sense to make the encrypted metadata an object carrying the information about selected encryption algorithm?

File: specifications/api/metadata-encrypt.md [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    {
+    ...
+      "encrypt_metadata":
+          { "passphrase": "metadata-secret"
+          },
+      "metadata":
+          { "1": "Hard times create strong men."
+          , "2": "Strong men create good times."
+          , "3": "Good times create weak men."
+          , "4": "And, weak men create hard times."
```
Comment (Unisay) [inline specifications/api/metadata-encrypt.md:82]: 💎

File: specifications/api/metadata-encrypt.md [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    ...
+      "metadata":
+         { "0": "0x0aa4f9a016217f75f10834367493f6d7e74197417ca25c7615cae02bc345382906fb6990daf8f138b2d9192e057d0d0b555f9d5fb287abb1842928c90f26e597"
+         , "1": "0x559ee85f00f1588b3ee32e81dc4c84aee208a10c1eec97fffe6e0e66c69d4e0b1e3e22d7edc1618df3b20b484527d86bc3bebad4295a2ad888d034b5fec38077"
+         , "2": "0x8d42154f681230124c64630ea68b841aec22f0530ec830cb662d59ef423ef23d7ff3"
+         }
+    ...
+    }
+    ```
+    as metadata values have 64-byte limit. In that case the encrypted metadata is encoded in the successive bytes.
```
Comment (Unisay) [inline specifications/api/metadata-encrypt.md:99]: So this makes it impossible to decrypt a single key without using the `cardano-wallet`?

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 8]
```diff
@@ -2619,19 +2621,26 @@ constructTransaction
...
-    body & \(ApiConstructTransactionData _ _ _ _ _ _ _ _) ->
+    body & \(ApiConstructTransactionData _ _ _ _ _ _ _ _ _) ->
     -- Above is the way to get a compiler error when number of fields changes,
     -- in order not to forget to update the pattern below:
         case body of
             ApiConstructTransactionData
                 { payments = Nothing
                 , withdrawal = Nothing
                 , metadata = Nothing
+                , encryptMetadata = Nothing
```
Comment (Unisay) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs:2632]: 👍🏻

**Review by Unisay — APPROVED** [review]

File: specifications/api/metadata-encrypt.md (hunk 5 above)
Comment (paweljakubas) [reply specifications/api/metadata-encrypt.md:41]: when we support other encryption algos this is going to be a must. but for now I made sure encrypt_metadata object could be extended in non-intrusive way for future. When we add it then we will add "string" with algo details along encrypted bytes. We will easily be able to handle this in noninvasive way and be backward compatible. thanks for the remark!

File: specifications/api/metadata-encrypt.md (hunk 7 above)
Comment (paweljakubas) [reply specifications/api/metadata-encrypt.md:99]: we need to append 64-byte chunks and then you can decrypt whole. And you can do it outside cardano-wallet if you comply with algos and passphrase symmetrizing algo

**Pull request merged**